File: component-vector/vector_search.js
#!/usr/bin/env node
'use strict';

/**
 * 变体向量检索模块（ES KNN）。
 *
 * 作为库：
 *   const { searchVariant } = require('./vector_search');
 *   const results = await searchVariant('主要按钮 悬停状态 小号', 3);
 *
 * CLI 测试：
 *   node vector_search.js "主要按钮 悬停状态 小号"
 *   MOCK_EMBED=1 node vector_search.js "文字链接 禁用"
 */

const fs   = require('fs');
const path = require('path');
const OpenAI = require('openai');
const { getClient } = require('./es_client');

// ── 配置 ─────────────────────────────────────────────────────────────────────

const envFile = path.resolve(__dirname, '.env');
if (fs.existsSync(envFile)) {
  fs.readFileSync(envFile, 'utf8').split('\n').forEach(line => {
    const t = line.trim();
    if (!t || t.startsWith('#')) return;
    const idx = t.indexOf('=');
    if (idx < 0) return;
    const k = t.slice(0, idx).trim();
    const v = t.slice(idx + 1).trim();
    if (k && !process.env[k]) process.env[k] = v;
  });
}

const MOCK_MODE          = /^(1|true|yes)$/i.test(process.env.MOCK_EMBED || '');
const EMBEDDING_BASE_URL = process.env.EMBEDDING_BASE_URL || 'https://dashscope.aliyuncs.com/compatible-mode/v1';
const EMBEDDING_MODEL    = process.env.EMBEDDING_MODEL    || 'text-embedding-v3';
const EMBEDDING_DIM      = Number(process.env.EMBEDDING_DIM) || 1024;
const ES_INDEX           = process.env.ES_INDEX || 'component_variants';

const openai = new OpenAI({
  apiKey:  process.env.DASHSCOPE_API_KEY,
  baseURL: EMBEDDING_BASE_URL,
  timeout: 30_000,
});

// ── Embedding ─────────────────────────────────────────────────────────────────

function mockEmbedding() {
  const v = Array.from({ length: EMBEDDING_DIM }, () => Math.random() * 2 - 1);
  const norm = Math.sqrt(v.reduce((s, x) => s + x * x, 0));
  return v.map(x => x / norm);
}

async function embedQuery(text) {
  if (MOCK_MODE) return mockEmbedding();
  const resp = await openai.embeddings.create({
    model:           EMBEDDING_MODEL,
    input:           [text],
    dimensions:      EMBEDDING_DIM,
    encoding_format: 'float',
  });
  return resp.data[0].embedding;
}

// ── 主检索函数 ────────────────────────────────────────────────────────────────

/**
 * @param {string} query  - 自然语言描述
 * @param {number} topK   - 返回条数，默认 1
 * @returns {Promise<Array<{score, symbol_id, variant_key, component_set_key, component_set_resolved, path}>>}
 */
async function searchVariant(query, topK = 1) {
  const es       = getClient();
  const queryVec = await embedQuery(query);

  const resp = await es.search({
    index: ES_INDEX,
    knn: {
      field:          'embedding',
      query_vector:   queryVec,
      k:              topK,
      num_candidates: Math.max(topK * 10, 50),
    },
    _source: ['symbol_id', 'variant_key', 'component_set_key', 'component_set_resolved', 'path', 'name', 'canvas_name', 'variant_name', 'text'],
    size: topK,
  });

  return resp.hits.hits.map(hit => ({
    score:                  hit._score,
    symbol_id:              hit._source.symbol_id,
    variant_key:            hit._source.variant_key,
    component_set_key:      hit._source.component_set_key,
    component_set_resolved: hit._source.component_set_resolved,
    path:                   hit._source.path,
    name:                   hit._source.name,
    canvas_name:            hit._source.canvas_name,
    variant_name:           hit._source.variant_name,
    _text:                  hit._source.text,
  }));
}

module.exports = { searchVariant };

// ── CLI ───────────────────────────────────────────────────────────────────────

if (require.main === module) {
  const query = process.argv[2];
  const topK  = Number(process.argv[3] || 5);

  if (!query) {
    console.error('用法: node vector_search.js "<查询>" [topK]');
    process.exit(1);
  }

  (async () => {
    console.log(`查询: "${query}"  topK=${topK}  ${MOCK_MODE ? '[MOCK]' : `[${EMBEDDING_MODEL}]`}\n`);
    const results = await searchVariant(query, topK);

    results.forEach((r, i) => {
      console.log(`#${i + 1}  score=${r.score?.toFixed(4)}`);
      console.log(`  名称:               ${r.name}  [${r.canvas_name}]`);
      console.log(`  变体:               ${r.variant_name || '(无变体)'}`);
      console.log(`  文本:               ${r._text}`);
      console.log(`  symbol_id:          ${r.symbol_id}`);
      console.log(`  variant_key:        ${r.variant_key}`);
      console.log(`  component_set_key:  ${r.component_set_key}`);
      console.log(`  path:               ${r.path}`);
      console.log();
    });
  })().catch(err => { console.error(err.message); process.exit(1); });
}

File: component-vector/vector_search.py
#!/usr/bin/env python3
"""
组件变体向量检索模块。

作为库：
  from vector_search import search_variant, search_variant_batch
  results = search_variant('主要按钮悬停状态', top_k=3)

CLI：
  python vector_search.py "主要按钮悬停" 3
  MOCK_EMBED=1 python vector_search.py "文字链接 禁用"
"""

import sys
from typing import Dict, List
from config import ES_INDEX, MOCK_MODE, EMBEDDING_MODEL
from es_client import get_client
from embed_client import embed_many

_SOURCE = [
    'symbol_id', 'variant_key', 'component_set_key', 'component_set_resolved',
    'path', 'name', 'canvas_name', 'variant_name', 'text',
]


def _hit_to_dict(hit: dict) -> dict:
    src = hit['_source']
    return {
        'score':                  hit['_score'],
        'symbol_id':              src.get('symbol_id'),
        'variant_key':            src.get('variant_key'),
        'component_set_key':      src.get('component_set_key'),
        'component_set_resolved': src.get('component_set_resolved'),
        'path':                   src.get('path'),
        'name':                   src.get('name'),
        'canvas_name':            src.get('canvas_name'),
        'variant_name':           src.get('variant_name'),
        '_text':                  src.get('text'),
    }


def search_variant(query: str, top_k: int = 1) -> List[Dict]:
    es  = get_client()
    vec = embed_many([query])[0]

    resp = es.search(
        index=ES_INDEX,
        knn={
            'field':          'embedding',
            'query_vector':   vec,
            'k':              top_k,
            'num_candidates': max(top_k * 10, 50),
        },
        source=_SOURCE,
        size=top_k,
    )
    return [_hit_to_dict(h) for h in resp['hits']['hits']]


def search_variant_batch(queries: List[str], top_k: int = 1) -> List[List[Dict]]:
    """每个 query 独立返回 top_k 结果，返回与 queries 等长的二维列表。"""
    if not queries:
        return []
    es   = get_client()
    vecs = embed_many(queries)

    searches = []
    for vec in vecs:
        searches.append({'index': ES_INDEX})
        searches.append({
            'knn':    {'field': 'embedding', 'query_vector': vec, 'k': top_k, 'num_candidates': max(top_k * 10, 50)},
            'source': _SOURCE,
            'size':   top_k,
        })

    resp = es.msearch(searches=searches)
    return [
        [_hit_to_dict(h) for h in r['hits']['hits']]
        for r in resp['responses']
    ]


if __name__ == '__main__':
    query = sys.argv[1] if len(sys.argv) > 1 else None
    top_k = int(sys.argv[2]) if len(sys.argv) > 2 else 5

    if not query:
        print('用法: python vector_search.py "<查询>" [top_k]', file=sys.stderr)
        sys.exit(1)

    label = '[MOCK]' if MOCK_MODE else f'[{EMBEDDING_MODEL}]'
    print(f'查询: "{query}"  top_k={top_k}  {label}\n')

    for i, r in enumerate(search_variant(query, top_k)):
        print(f'#{i + 1}  score={r["score"]:.4f}')
        print(f'  名称:               {r["name"]}  [{r["canvas_name"]}]')
        print(f'  变体:               {r["variant_name"] or "(无变体)"}')
        print(f'  文本:               {r["_text"]}')
        print(f'  symbol_id:          {r["symbol_id"]}')
        print(f'  variant_key:        {r["variant_key"]}')
        print(f'  component_set_key:  {r["component_set_key"]}')
        print(f'  path:               {r["path"]}')
        print()
